dimensiondata/src/main/java/org/jclouds/dimensiondata/cloudcontrol/features/ServerApi.java

```diff
@@ -0,0 +1,216 @@
+import org.jclouds.rest.annotations.PayloadParam;
+import org.jclouds.rest.annotations.RequestFilters;
+import org.jclouds.rest.annotations.ResponseParser;
+import org.jclouds.rest.annotations.Transform;
+import org.jclouds.rest.binders.BindToJsonPayload;
+
+import javax.inject.Inject;
+import javax.inject.Named;
+import javax.inject.Singleton;
+import javax.ws.rs.Consumes;
+import javax.ws.rs.GET;
+import javax.ws.rs.POST;
+import javax.ws.rs.Path;
+import javax.ws.rs.PathParam;
+import javax.ws.rs.Produces;
+import javax.ws.rs.core.MediaType;
+import java.util.List;
+
+@RequestFilters({ BasicAuthentication.class, OrganisationIdFilter.class })
+@Consumes(MediaType.APPLICATION_JSON)
+@Path("/{jclouds.api-version}/server")
+public interface ServerApi {
+
+   @Named("server:list")
+   @GET
+   @Path("/server")
+   @ResponseParser(ParseServers.class)
+   @Fallback(Fallbacks.EmptyIterableWithMarkerOnNotFoundOr404.class)
+   PaginatedCollection<Server> listServers(PaginationOptions options);
+
+   @Named("server:list")
+   @GET
+   @Path("/server")
+   @Transform(ParseServers.ToPagedIterable.class)
+   @ResponseParser(ParseServers.class)
+   @Fallback(Fallbacks.EmptyPagedIterableOnNotFoundOr404.class)
+   @RequestFilters({ DatacenterIdFilter.class })
+   PagedIterable<Server> listServers();
+
+   @Named("server:get")
+   @GET
+   @Path("/server/{id}")
+   @Fallback(Fallbacks.NullOnNotFoundOr404.class)
+   Server getServer(@PathParam("id") String id);
+
+   @Named("server:deploy")
+   @POST
+   @Path("/deployServer")
+   @Produces(MediaType.APPLICATION_JSON)
+   @MapBinder(BindToJsonPayload.class)
+   @ResponseParser(ServerId.class)
+   String deployServer(@PayloadParam("name") String name, @PayloadParam("imageId") String imageId,
+         @PayloadParam("start") Boolean start, @PayloadParam("networkInfo") NetworkInfo networkInfo,
+         @PayloadParam("administratorPassword") String administratorPassword);
+
+   @Named("server:deploy")
+   @POST
+   @Path("/deployServer")
+   @Produces(MediaType.APPLICATION_JSON)
+   @MapBinder(CreateServerOptions.class)
+   @ResponseParser(ServerId.class)
+   String deployServer(@PayloadParam("name") String name, @PayloadParam("imageId") String imageId,
+         @PayloadParam("start") Boolean start, @PayloadParam("networkInfo") NetworkInfo networkInfo,
+         @Nullable @PayloadParam("administratorPassword") String administratorPassword,
+         @Nullable @PayloadParam("disk") List<Disk> disks, @Nullable CreateServerOptions options);
+
+   @Named("server:delete")
+   @POST
+   @Path("/deleteServer")
+   @Produces(MediaType.APPLICATION_JSON)
+   @Fallback(Fallbacks.VoidOnNotFoundOr404.class)
+   @MapBinder(BindToJsonPayload.class)
+   void deleteServer(@PayloadParam("id") String id);
+
+   @Named("server:powerOff")
+   @POST
+   @Path("/powerOffServer")
+   @Produces(MediaType.APPLICATION_JSON)
+   @MapBinder(BindToJsonPayload.class)
+   @Fallback(Fallbacks.VoidOnNotFoundOr404.class)
```

Comment:
In state change operations I think it would be better to fail if the server does not exist. Otherwise, users won't be able to differentiate a failed execution than a succeeded one. The use case here is different than in get operations or delete (where the purpose is to have the server gone). Let's remove the 404 fallbacks from all state change operations.

Reply:
What you are saying makes sense. I will remove this from the API definition.